### DISCUSSION TITLE
RP2040: Using picoprobe with a GUI/Editor like VSC or Sublime Text.

### DISCUSSION THREAD

**Discussion by robert-hh** [micropython/micropython#11028] [General]
picoprobe can nicely be used in combination with openocd and GDB to debug rp2040 code. But using GDB is not very comfortable. Has anyone been successful in using it e.g. with the GDB plugin for VSC or knows a more convenient GUI for GDB?

**Comment by robert-hh** [comment]
To answer my own question: there is `kdbg`, which works well. It may not be in the standard app library of your Linux version, so you have to compile it yourself. The repository is https://github.com/j6t/kdbg.git.
It is somewhat more convenient than using the command line GDB, even if the latte rmay have more options. But for most of my debugging needs it's fine. 
To work e.g. with rp2040, you obviously have to 
- start openocd first,
- set `arm-none-eabi-gdb` as debugger in the `Global Options` or  `This Program` options of kdbg.
- start kdbg with the command line: `kdbg -r localhost:3333 <elf-file-name>'. The elf-file-name is optional. You can as well load it from the kdbg main window.

Then you have the common Fxx keys for run, single step, setting breakpoints, watch variables, etc. Since it is GUI front end for GDB, much of that is only available when the program is halted, and for that you have to go through the menu, or you define your own hotkey. All keyboard hotkeys are fully configurable.

**Comment by cve2022** [comment]
Robert, what about [using VSCode for that](https://www.digikey.com/en/maker/projects/raspberry-pi-pico-and-rp2040-cc-part-2-debugging-with-vs-code/470abc7efb07432b82c95f6f67f184c0) ?

**Reply by robert-hh** [reply]
I have seen these instructions and I tried them, but could not get them to work. They may work for simple files loaded to the pico, but it failed with the firmware.elf package. Similar fails were my attempt to use a GBD add-on to sublime text, which I used as code editor. If you have a working configuration, Please share it with me.
So I ended up with kdbg, which works well and is fully configurable.

I seems that one can use the RPi picoprobe together with openocd as generic JTAG based SWD debugger. Quite a few architectures are supported, and compared to the Segger stuff it is pretty low price, like 4€ for a Pico or 12€ for the form factor as Debug probe, which is small and comes in a pretty plastic case.

**Reply by cve2022** [reply]
Robert, I should have guessed you tried this. I am not developing in C/C++ with the Pico, so no configuration to share, unfortunately.

**Reply by robert-hh** [reply]
I do not worry. I'm happy now with kdbg. I would have used VSC only for that purpose, and for debugging VSC is not great, even for  a simple "Hello World" example. For such a complex project like MicroPython I never could get it to work as Debugger. For Code editing I switched to Sublime text. There is also a Plug-In for GDB, and I got it once to work somewhat, but the layout is poor. 
kdbg is a single purpose tool focused on debugging, and as such it works well. And it fits my habit for using dedicated tools for dedicated tasks.

**Reply by cve2022** [reply]
FWIW, in another project where I use the STM32, I use VSCode for their editor capabilities and Keil for debugging. I will take a look at Sublime though.
